.github/workflows/lint.yml

```diff
@@ -0,0 +1,131 @@
+name: XGBoost CI (Lint)
+
+on: [push, pull_request]
+
+permissions:
+  contents: read  # to fetch code (actions/checkout)
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+env:
+  BRANCH_NAME: >-
+    ${{ github.event.pull_request.number && 'PR-' }}${{ github.event.pull_request.number || github.ref_name }}
+
+jobs:
+  build-containers:
+    name: Build CI containers
+    env:
+      CONTAINER_ID: xgb-ci.clang_tidy
+    runs-on:
+      - runs-on=${{ github.run_id }}
+      - runner=linux-amd64-cpu
+    steps:
+      # Restart Docker daemon so that it recognizes the ephemeral disks
+      - run: sudo systemctl restart docker
+      - uses: actions/checkout@v4.2.2
+        with:
+          submodules: "true"
+      - name: Build ${{ env.CONTAINER_ID }}
+        run: bash ops/docker_build.sh
+
+  clang-tidy:
+    name: Run clang-tidy
+    needs: build-containers
+    runs-on:
+      - runs-on=${{ github.run_id }}
+      - runner=linux-amd64-cpu
+    steps:
+      # Restart Docker daemon so that it recognizes the ephemeral disks
+      - run: sudo systemctl restart docker
+      - uses: actions/checkout@v4.2.2
+        with:
+          submodules: "true"
+      - name: Fetch container from cache
+        run: bash ops/docker_build.sh
+        env:
+          CONTAINER_ID: xgb-ci.clang_tidy
+      - run: bash ops/pipeline/run-clang-tidy.sh
+
+  python-mypy-lint:
+    runs-on: ubuntu-latest
+    name: Type and format checks for the Python package
+    steps:
+      - uses: actions/checkout@v4.2.2
+        with:
+          submodules: 'true'
+      - uses: conda-incubator/setup-miniconda@v3.1.0
+        with:
+          miniforge-variant: Miniforge3
+          miniforge-version: latest
```

Comment:
Since this gets repeated so much, you may want to define it as a repository environment variable: https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment

Try to minimize hard-coded values where you can - especially for things that are hard to do search/replace on, like "latest"